[framework]Add semantics role to table rows.

hannah-hyj opened the pull request:
**1. framework side:**
This PR Create semantics node for rows in table's `assembleSemanticsNode` function. 

**2. web side:**
I tested on my mac, and i need to remove the 
`<flt-semantics-container>` between table and row, row and cell to traverse inside table, removing those transfom intermediate containers on web will be a bit of hassle and will be in another separate PR.

For example this code can only announce table but can’t get into cells.
```
<flt-semantics id="flt-semantic-node-4" role="table" style="position: absolute; overflow: visible; width: 751px; height: 56px; transform-origin: 0px 0px 0px; transform: matrix(1, 0, 0, 1, 0, 56); pointer-events: none; z-index: 1;">
    <flt-semantics-container style="position: absolute; pointer-events: none; top: 0px; left: 0px;">
        <flt-semantics id="flt-semantic-node-6" role="row" style="position: absolute; overflow: visible; width: 751px; height: 56px; top: 0px; left: 0px; pointer-events: none;">
            <flt-semantics-container style="position: absolute; pointer-events: none; top: 0px; left: 0px;">
                <flt-semantics id="flt-semantic-node-5" role="columnheader" aria-label="Name" style="position: absolute; overflow: visible; width: 751px; height: 56px; top: 0px; left: 0px; pointer-events: all;"></flt-semantics>
            </flt-semantics-container>
        </flt-semantics>
    </flt-semantics-container>
</flt-semantics>
```
If I removed the in between `</flt-semantics-container>`, the code come 
```
<flt-semantics id="flt-semantic-node-4" role="table" style="position: absolute; overflow: visible; width: 751px; height: 56px; transform-origin: 0px 0px 0px; transform: matrix(1, 0, 0, 1, 0, 56); pointer-events: none; z-index: 1;">
    <flt-semantics id="flt-semantic-node-6" role="row" style="position: absolute; overflow: visible; width: 751px; height: 56px; top: 0px; left: 0px; pointer-events: none;">
        <flt-semantics id="flt-semantic-node-5" role="columnheader" aria-label="Name" style="position: absolute; overflow: visible; width: 751px; height: 56px; top: 0px; left: 0px; pointer-events: all;"></flt-semantics>
    </flt-semantics>
</flt-semantics>
```
And I can get into table cells.


**3. Other aria-attributes:** 
[aria-colcount](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-colcount) ,[aria-rowcount](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-rowcount)  [aria-colindex](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-colindex)  [aria-rowindex](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-rowindex)  
* theres attributes are only needed if some rows and columns are hidden in the Dom tree.  havn't added them yet

aria-rowspan , aria-colspan : 
*we currently don't support row span and col span in our widgets. related issue: https://github.com/flutter/flutter/issues/21594

 related: https://github.com/flutter/flutter/pull/162339
issue: https://github.com/flutter/flutter/issues/45205

## Pre-launch Checklist

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [ ] I signed the [CLA].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

hannah-hyj commented:
hmmm looks like i should not simply remove <flt-semantics-container>, it will mess up with some transform. i will mark this PR as draft and do more testing.

yjbanov commented:
> hmmm looks like i should not simply remove `<flt-semantics-container>`, it will mess up with some transform. i will mark this PR as draft and do more testing.

Oof, that's a tricky problem. What's the typical value of `rect` passed for tables in semantic updates. I wonder if they are simple enough that we could use an alternative strategy that doesn't need auxiliary child container elements. For example, I think if `rect.top` and `rect.left` are zero, we can do everything with just one element.

hannah-hyj committed:
Update table.dart

Update semantics.dart

Update table.dart

assembleSemanticsNode

1

hannah-hyj committed:
lint

hannah-hyj committed:
Add comments

hannah-hyj committed:
lint

hannah-hyj commented:
@chunhtai  I want to merge this  framework changes PR first and the "removing intermediate container layers in web engine" will be a separate effort in another PR.

hannah-hyj committed:
Update data_table_test.dart

hannah-hyj committed:
fix tests

hannah-hyj committed:
Update table.dart

hannah-hyj committed:
Update table_test.dart

chunhtai commented on packages/flutter/lib/src/rendering/table.dart:
```diff
@@ -607,9 +607,57 @@ class RenderTable extends RenderBox {
+
+      final SemanticsConfiguration configuration =
+          SemanticsConfiguration()
+            ..indexInParent = i
```
is this going to be used by the row index? if so, does the cell also uses this to set the col index? we should unify the way to set these indices in web.

hannah-hyj replied:
updated indexInParent for cells

chunhtai commented on packages/flutter/test/widgets/table_test.dart:
```diff
@@ -84,11 +85,14 @@ void main() {
-                  children: const <TableRow>[
-                    TableRow(children: <Widget>[Text('AAAAAA'), Text('B'), Text('C')]),
-                  ],
+                TableCell(
```
Is this needed for semantics check? do we need to worry about this being a breaking change?

hannah-hyj replied:
yes because the test case here before is a very corner case putting a table inside a row, so it became  table-> row-> table and it failed the new check, i think cases like this are very rare and they are just made up for tests

chunhtai replied:
no, this is a case i replicate from internal code.

Should the tableRow assign the TableCell to its children internally?

hannah-hyj replied:
Should the tableRow assign the TableCell to its children internally? 
I'm already doing that in assembleSemanticsNode, i assign the role to its children if the children have no roles.
but if the children already have roles(in this case, it's table), do we want to override that role? or do we want to insert a semantics node with a cell role between them?

chunhtai replied:
> I'm already doing that in assembleSemanticsNode

Assigning child property in assembleSemanticsNode seems a bit hacky. We should do it through the TableCell widget if possible.

> do we want to insert a semantics node with a cell role between them

This would be ideal, but doing it in widget layer means the Table here needs to be wrap with a semantics container. I think for now, let's just override it without creating a container. As you said this may be a rare case, we can fix it if it really is an issue

chunhtai replied:
also looks like this pr will make the table to be a container so there may not be overriding role issue in this particular case

hannah-hyj replied:
>Assigning child property in assembleSemanticsNode seems a bit hacky. We should do it through the TableCell widget if possible.
We can't because we don't enforce user to use `TableCell` widget in table, TableRow can take any widgets as its children. User can just assign a Text widget to TableRow's children.  That's why I assign the role here. 

>I think for now, let's just override it without creating a container.
SGTM,
 
>there may not be overriding role issue in this particular case 
in This particular case, the test case was table-> row-> table- > row-> cell, If I don't override the second table role , it will fail the check because row's children shouldn't be a table, if i override the second table role to be a cell, it will become table-> row-> cell- > row-> cell, it will fail the check because row's parent shouldn't be a cell. So I see the only way to fix this test case is to insert a cell and it will become table-> row-> cell -> table- > row-> cell,

hannah-hyj replied:
>This would be ideal

Update this PR to insert a semantics node with a cell role between them.

hannah-hyj committed:
shift transform instead of rect

hannah-hyj committed:
update indexInParent

hannah-hyj commented:
I just realized I had some misunderstanding about semanticsNode rect before.   I thought the tree dump rect is actual semanticsnode.rect, but it's actually semanticsnode.rect with semanticsnode.transform.

Instead of only setting  row's and cell's `rect`, actually i should set their `transform`.  Because `rect` in the a semantics node is something that won't change their children's position, only `transform` will. 

Updated PR.

hannah-hyj committed:
update rects to be more clear

hannah-hyj committed:
update the loop

Update table.dart

Update table.dart

lint

Update table.dart

fix a test

chunhtai commented on packages/flutter/lib/src/semantics/semantics.dart:
```diff
@@ -2869,6 +2895,13 @@ class SemanticsNode with DiagnosticableTreeMixin {
   /// {@endtemplate}
   SemanticsRole get role => _role;
   SemanticsRole _role = _kEmptyConfig.role;
+  set role(SemanticsRole value) {
```
we should avoid this

hannah-hyj replied:
This is removed after offline discussion.

we won't directly change semantics node's role, just wrap them with a new semantics node with a cell role.

chunhtai commented:
> We can't because we don't enforce user to use TableCell widget in table

I mean we wrap it internally in either Table widget or TableRow widget

chunhtai commented:
> I mean we wrap it internally in either Table widget or TableRow widget

looking at the code, we shouldn't add semanticsRole.cell in TableCell. It should added internally in Table or TableRow, this also avoid forcing people to use TableCell to have the correct semantics

hannah-hyj commented:
I mean we wrap it internally in either Table widget or TableRow widget>> 
If you mean like to wrap TableRow's every children with Semantics( role: cell: child: child), I tried that and it didn't work out because TableCell is a ParentDataWidget so it assumes its parent is a Table, insert a Semantics widget between TableCell and Table will crash it. So I chose to assign the semantics role to cells in assembleSemanticsNode although it looks a bit hack 😅  maybe there is a better way to do it ?   

we shouldn't add semanticsRole.cell in TableCell>> make sense, I  will remove that.

chunhtai commented:
this is unfortunate, I can't think of better way of doing this. I think this is fine then

hannah-hyj committed:
add comments

hannah-hyj committed:
test

hannah-hyj committed:
insert a cell role node  if it has existing roles

hannah-hyj committed:
update wrapper

hannah-hyj committed:
update if logic to fix scrolling cases.

hannah-hyj committed:
update comments

hannah-hyj committed:
update comments

hannah-hyj committed:
update comments

hannah-hyj commented on packages/flutter/lib/src/rendering/table.dart:
```diff
@@ -607,9 +607,120 @@ class RenderTable extends RenderBox {
+
+        // Assign the role cell to the child if it doesn't have a role.
+        if (child.role == SemanticsRole.none) {
+          child.role = SemanticsRole.cell;
```
I tried and it worked correctly 😃 

code:
```
                TableRow(children: [
                  Semantics(
                    container: true,
                    child: Semantics(
                      role:
                          _flipped ? SemanticsRole.dialog : SemanticsRole.none,
                      child: Container(
                        padding: EdgeInsets.all(16.0),
                        child: Text('Cell with Semantics'),
                      ),
                    ),
                  ),
                  TableCell(
                      child: Padding(
                    padding: const EdgeInsets.all(16.0),
                    child: Text('Regular Cell'),
                  )),
                ]),
```
```
       ├─SemanticsNode#4
       │ │ Rect.fromLTRB(0.0, 428.5, 1137.0, 480.5)
       │ │ role: table
       │ │
       │ └─SemanticsNode#15
       │   │ Rect.fromLTRB(0.0, 0.0, 1137.0, 52.0)
       │   │ role: row
       │   │ indexInParent: 0
       │   │
       │   ├─SemanticsNode#5
       │   │   Rect.fromLTRB(0.0, 0.0, 568.5, 52.0)
       │   │   label: "Cell with Semantics"
       │   │   textDirection: ltr
       │   │   role: cell
       │   │   indexInParent: 0
       │   │
       │   └─SemanticsNode#6
       │       Rect.fromLTRB(568.5, 0.0, 1137.0, 52.0)
       │       label: "Regular Cell"
       │       textDirection: ltr
       │       role: cell
       │       indexInParent: 1
```

```
       ├─SemanticsNode#4
       │ │ Rect.fromLTRB(0.0, 428.5, 1137.0, 480.5)
       │ │ role: table
       │ │
       │ └─SemanticsNode#13
       │   │ Rect.fromLTRB(0.0, 0.0, 1137.0, 52.0)
       │   │ role: row
       │   │ indexInParent: 0
       │   │
       │   ├─SemanticsNode#14
       │   │ │ Rect.fromLTRB(0.0, 0.0, 568.5, 52.0)
       │   │ │ role: cell
       │   │ │ indexInParent: 0
       │   │ │
       │   │ └─SemanticsNode#5
       │   │     Rect.fromLTRB(0.0, 0.0, 568.5, 52.0)
       │   │     label: "Cell with Semantics"
       │   │     textDirection: ltr
       │   │     role: dialog
       │   │
       │   └─SemanticsNode#6
       │       Rect.fromLTRB(568.5, 0.0, 1137.0, 52.0)
       │       label: "Regular Cell"
       │       textDirection: ltr
       │       role: cell
       │       indexInParent: 1
```

hannah-hyj commented on packages/flutter/lib/src/rendering/table.dart:
```diff
@@ -607,9 +607,120 @@ class RenderTable extends RenderBox {
+        // Otherwise if the cell index in parent is not null, it means the cell's parent is set to a row,
+        // check its index to see if it belongs to the current row or next row.
+        if ((child.indexInParent == null && childRect.top > rowBox.bottom) ||
+            (child.indexInParent != null && child.indexInParent != x)) {
```
yes, it is possible.

chunhtai replied:
If this is stale should we worry about setting the wrong row because of it?

hannah-hyj replied:
should we worry about setting the wrong row because of it?>> 

In my local testing i tried scrolling and change browser size, with those changes, the cell will get stale, but their index in parent wont change, so i have no problem setting rows to them.

hannah-hyj replied:
stale-> dirty

hannah-hyj commented on packages/flutter/lib/src/rendering/table.dart:
```diff
@@ -607,9 +607,120 @@ class RenderTable extends RenderBox {
+        cell.indexInParent = x;
+
+        // Shift the cell's transform to be relative to the row.
+        final Matrix4 cellTransform = Matrix4.translationValues(_columnLefts!.elementAt(x), 0, 0);
```
I think because the cell here can be stale, and its parent might be a row or a table at this point, and the cell's transform and rect is relative (row or table?), so it's hard to get them from the cell.

chunhtai replied:
The rect and transform at this point should be accurate in relative to table. in order to call this method, the children's geometry must have be updated. are you seeing otherwise?

hannah-hyj replied:
Yes, it appears when i scroll, the cell can be stale here and their rect and transform is relative to row at this point.

I think it's because I changed their parent in last `assembleSemantics`,

chunhtai replied:
This is surprising, maybe a corner case in the semantics compiling. I think this is fine for now.

hannah-hyj committed:
resolve some comments

chunhtai commented:
left two comments https://github.com/flutter/flutter/pull/163337#discussion_r1984159363 and https://github.com/flutter/flutter/pull/163337#discussion_r1984161385

hannah-hyj committed:
resolve comments

hannah-hyj committed:
lint

hannah-hyj commented:
@chunhtai  PR updated after offline discussion

hannah-hyj committed:
change TableCell constructor back to a const one to avoid breaking change

hannah-hyj committed:
add a eps

add a eps

add a eps

lint

lint

chunhtai reviewed (approved):
LGTM

hannah-hyj committed:
support multiple semantics nodes inside one cell

hannah-hyj committed:
simplify code

hannah-hyj committed:
update to fix g3 tests

use a local map to record index

update transform

Update semantics.dart

add a if

add more tags

hannah-hyj committed:
update transform logic

hannah-hyj committed:
remove  container: true in _TableCell

hannah-hyj committed:
Update table.dart

hannah-hyj committed:
add boundary check

hannah-hyj committed:
columnheader can be inside cell wrapper

hannah-hyj committed:
cache rows and cells node

hannah-hyj committed:
Merge branch 'master' into table-role-2
